Plot changes: Step 1 A- Draft API for backend classes (#1382)
sherpa/sherpa: docs/plots/backends.rst

************************
Plotting backend details
************************

.. warning::
   The plotting backend API is currently re-designed and details might change in the near future.

Plotting in Sherpa is done through *plotting backends*. A *backend* is an

> external package that performs the actual plotting function, i.e. it creates a

Reviewer: you can add example of a backend in this sentence - 'for example matplotlib' package.

canvas and puts color on paper or on the screen. The Sherpa plotting classes are
written in a backend-independent way; they work with any plotting backend that
is supported by Sherpa. We also use the term *backend* to refer to Sherpa

> classes that connect the Sherpy plot objects with the plotting backend by

Reviewer: typo - Sherpa plot objects

translating Sherpa options into backend-specific commands. The details of that
are explained below, but, in short, a backend class would translate the sherpa
backend-independent "plot a line in red from here to there" into backend
specific ``plt.plot(x,y, linecolor='r')`` (if the backend is :term:`matplotlib`).

Instead of utilizing Sherpa classes and commands to plot data, users can
alternatively just access the data in the Sherpa data objects (e.g. the x and y
values of a dataset) and perform plotting operations directly with any plotting

> backend that is available ot them. This method may be less convenient, but it

Reviewer 2: ```suggestion
backend that is available to them. This method may be less convenient, but it
```

works for ploting backends not (yet) supported by Sherpa.

Backend-independent plotting options
====================================

Sherpa defines a number of plotting options that can be used with any backend,
thus code that limits itself to those options can run independent of which
backend is set up in Sherpa. All default settings and most examples use only
backend-inpdependent plotting options. The resulting plots will not look
identical in each backend, e.g. thickness of a line or the font type of an
annotation might differ, but they will convey the same information. For example,
"a blue dotted line" will generally appear blue and dotted, even though the
shade of blue or the size of the dots might differ between plotting backends. 

In some cases, a plotting backend might not support all Sherpa plot options (for
example, a plotting backend might not have a command to change the line style).
In those rare cases, a setting might be ignored; but a setting from Sherpa's

> list of backend independent value will never raise an error.

Reviewer 2: ```suggestion
list of backend independent values will never raise an error.
```

The names of plotting options and values that all Sherpa plotting backends
accept are chosen to match common :term:`matplotlib` settings which are familiar
to many scientific Python users, and to maintain backwards compatibility with
previous version of Sherpa. They offer limited choices, but those are sufficient

> for most plots. For example, only few different line styles are specified, but

Reviewer 2: ```suggestion
for most plots. For example, only a few different line styles are specified, but
```

> in practice, most plots can be done with just solild, dashed, and maybe dotted

Reviewer 2: ```suggestion
in practice, most plots can be done with just solid, dashed, and maybe dotted
```

lines.

The following settings are accepted for all Sherpa plotting backends:

Colors
------

Colors are used for the color of lines, symbols, errorbars etc.

- ``'b'`` (blue)
- ``'r'`` (red)
- ``'g'`` (green)
- ``'k'`` (black)
- ``'w'`` (white), 
- ``'c'`` (cyan)
- ``'y'`` (yellow)
- ``'m``` (magenta)

> - ``None`` (ploting backend default)

Reviewer 2: ```suggestion
- ``None`` (plotting backend default)
```

Most plotting backends accept more than just the backend-independent options
listed above. For example, :term:`matplotlib` allows `many different ways to
specify colors to cover the entire RGB range
<https://matplotlib.org/stable/tutorials/colors/colors.html>`_, such as
``color=(.3, .4, .5, .2)`` or ``color='xkcd:eggshell'``. The Sherpa plotting
methods will pass any value to the underlying backend, but only the options in
the backend independent list are guaranteed to work in every case. For example, 

  >>> from sherpa.data import Data1D
  >>> from sherpa.plot import DataPlot
  >>> d = Data1D('example data', [1, 2, 3], [3, 2, 5])
  >>> dplot = DataPlot()
  >>> dplot.prepare(d)
  >>> dplot.plot(markerfacecolor='xkcd:eggshell')

will succeed with the `~sherpa.plot.pylab_backend.PylabBackend`, but raise an
error if the active backend is `~sherpa.plot.bokeh_backend.BokehBackend`. In
contrast, ``color='k'`` is also not understood by bokeh natively, but because it is on
the backend-independent list, Sherpa will translate ``'k'`` to a form that bokeh
does understand (``'black'`` in this case).

Backends may also accept additional keywords to specify more plotting properties

> such as the transparancy of an element or an URL that is opened when clicking on

Reviewer: typo in 'transparency'

In some cases, the Sherpa plotting commands create several visualization
elements at the same time (lines, symbols, error bars, axes, labels). This makes
using Sherpa classes convenient, but it also means that the plotting functions
do not offer options to customize each and every part. In general, the plotting
functions pass color, linestyle etc. to the elements that describes the data
(line, marker) and generate labels or axes grids using default settings. Backend
specific code can be used to change the properties of the current figure after
the Sherpa plotting.

Backend interface
=================

.. note::

   This section is mostly relevant for developers or advanced users who write new
   Sherpa plot classes or new backends.

This section describes the API that all Sherpa backends offer to explain how to
use it and why it was designed this way. See `sherpa.plot.backend.BaseBackend`
for a complete listing of the calling signature for each function. 
The `sherpa.plot.backend.BasicBackend` backend extends 
`sherpa.plot.backend.BaseBackend` by raising a warning message for 
plotting functions (plot, image, histrogram etc.) that are not implemented.

> It is a the base for any real functional backend, which will override those

Reviewer 2: ```suggestion
It is the base for any real functional backend, which will override those
```

does not support 2D data. In that case, the backend would typically issue a
warning.

The plotting functions are not separated by "how things look on paper" (thus "plot" is
a long method that is responsible for points, lines, and errorbars), but
by "what is the input data type":

- `~sherpa.plot.backend.BaseBackend.plot` (for scatter plots with markerstyle
  set, for line plots with linestyle set, and for errorbars with ``xerr`` or
  ``yerr`` set to `True`); accepts (x, y) data with optional error bars in each
  dimension. Data can be scalar (for a single marker), or array-like.
- `~sherpa.plot.backend.BaseBackend.histo` (similar to plot, but with
  "histogram-style" lines); accepts (xlo, xhi, y) data with optional xerr, yerr.
- `~sherpa.plot.backend.BaseBackend.contour` for (x0, x1, z) data
- `~sherpa.plot.backend.BaseBackend.image` for (x0, x1, z) data on a regular
  grid. An image is different from a contour in the sense that an image is
  pixelated on a regular grid, while a contour can in principle describe a
  continuous quantity or an irregular grid, even if the current implementation
  may not provide that flexibility.

Annotations
-----------

> Backends should also implement the follwing annotation functions. They do not

Reviewer 2: ```suggestion
Backends should also implement the following annotation functions. They do not
```

depend on the data plotted, but just annotate the plot, e.g. a

> `~sherpa.plot.RatioPlot` shows the ratio betwen data and model and can use an

Reviewer 2: ```suggestion
`~sherpa.plot.RatioPlot` shows the ratio between data and model and can use an
```

annotation to mark the ``ratio=1`` line.

- `~sherpa.plot.backend.BaseBackend.hline` (horizontal across the entire axes)
- `~sherpa.plot.backend.BaseBackend.vline` (vertical across the entire axes)

Other annotations (e.g. text labels) might be added to the API in the future.
For this reason new backends should inherit from
`~sherpa.plot.backend.BasicBackend`. Any function added to the API will be
implemented in `~sherpa.plot.backend.BasicBackend` as a no-op with a
warning to the user like "Feature XYZ is not available in your backend ". That

> way, all Shepa plots can immediately make use of newly added functions without

Reviewer 2: ```suggestion
way, all Sherpa plots can immediately make use of newly added functions without
```

breaking existing plotting backends; the worst that happens is that not all
annotation will be visible in every backend.

Return values
-------------

Sherpa does not expect a specific return argument from any plotting function,
but they are allowed to have return values if that is helpful for their internal
implementation, e.g. in the matplotlib backend, plotting a line might return a
line object so that error bars plotted later can use ``line.color`` to match the
color of that line.

Creating plots and panels, clearing and overplotting
----------------------------------------------------
At this stage, we keep the existing API for creating plot and panels, for
clearing and overplotting, i.e. each of the plotting functions above accepts
 the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow

Multi-panels plot can be set with clear_window, set_subplot, set_jointplot
[copy documentation from CAIO 4.14 pylab_backend.clear_window etc. into here because that describes use and function]

> 

Reviewer: Do you include the limits on different axis when you create a plot? or the limits are somehow autogenerated?

Reviewer 3: I would be very surprised if limits can be set here, as that would be a big change in functionality (not necessarily a bad one, just not one I'd expect here).

Also, CIAO is typo-d here.

Author: In this PR, I try ot keep any changes to the API as minimal as possible. Limits are created as before (i.e. whatever the plotting backend chooses, typically the next round value below and above the mean) until the user resets it, see. e.g. https://cxc.cfa.harvard.edu/sherpa/threads/setplot_manual/ for instructions. We can have a separate issue if we want to add parameters for the limits, just to confirm what @DougBurke said above: Adding new keywords for the limits is out of scope for this PR.

Interaction with interactive plots in the UI
--------------------------------------------
Each backend has additional functions that are called before, during and after
interactive plots (begin, exception, and end), and for the setup of 
multi-panel plots [those are taken essentially unchanged from the 4.14 version, so can copy from code into specs]

Other methods
--------------

Backends need to have a few methods

- ``as_html_XXX`` (where XXX is a plot type) that are used for interactive
  display in the notebook with ``_repr_html_``.  These functions take a plot
  object and return an html representation as a string.
- ``get_XXX_plot/hist_prefs`` (where XXX is a plot type) which returns a
  dictionary of preferences that is used for displaying this plot.
- `~sherpa.plot.backend.BasicBackend.get_latex_for_string` to format latex in strings.

> How do I implements a new backend?

Reviewer 2: ```suggestion
How do I implement a new backend?
```

==================================

Example for a modified backend
------------------------------
Since Sherpa backends are defined using inheritance, new backends can be
created to change the appearance of plots in ways that are beyond the

> options offered as keword arguments for the existing backends. As an example,

Reviewer 2: ```suggestion
options offered as keyword arguments for the existing backends. As an example,
```

    `sherpa.plot.backends.IndepOnlyBackend` does not perform any plotting,
    but it does raise a warning when plotting options that are not on the list of
    backend-independent parameters are used. Users will never use this backend,
    but running our tests with it ensures that we don't hardcode matplotlib-specific
    defaults into any of our objects. In principle, most plotting tests fall into this
    category, since most code needs to work for all backends. However, running every
    plotting test with every possible backend can make the runtime for the tests long.
    In practice, we thus do not run every test that we could for all backends.
    To run a test for all backends, add the ``all_plot_backends`` 
    `pytest fixture <https://docs.pytest.org/en/stable/explanation/fixtures.html>`_::

      def test_this_plotting_feature(all_plot_backends):
          ... test code here ...

    This fixture will execute the test several times with different active backends.
    Alternatively, sherpa also has a fixture that runs all backends *except* for the 
    `sherpa.plot.backends.IndepOnlyBackend`: `plot_backends`. This is useful if a test

>     makes use of options that not on the Backend-independent list and needs to avoid the

Reviewer 2: ```suggestion
    makes use of options that are not on the Backend-independent list and needs to avoid the
```

************************
Plotting backend details
************************

.. warning::
   The plotting backend API is currently re-designed and details might change in the near future.

Plotting in Sherpa is done through *plotting backends*. A *backend* is an
external package that performs the actual plotting function such as :term:`matplotlib`, i.e. it creates a
canvas and puts color on paper or on the screen. The Sherpa plotting classes are
written in a backend-independent way; they work with any plotting backend that
is supported by Sherpa. We also use the term *backend* to refer to Sherpa
classes that connect the Sherpa plot objects with the plotting backend by
translating Sherpa options into backend-specific commands. The details of that
are explained below, but, in short, a backend class would translate the sherpa
backend-independent "plot a line in red from here to there" into backend
specific ``plt.plot(x,y, linecolor='r')`` (if the backend is :term:`matplotlib`).

> 

Reviewer 3: I think this needs re-writing as you point out in the middle that we actually have two concepts of `backend` - the code we have written to interface to the backend, and the actual backend. I don't have the energy to write a suggestion just let but let me come back to it (unless anyone else has a strong desire to write something)

Author: I'll reword that.

Instead of utilizing Sherpa classes and commands to plot data, users can
alternatively just access the data in the Sherpa data objects (e.g. the x and y
values of a dataset) and perform plotting operations directly with any plotting
backend that is available to them. This method may be less convenient, but it
works for ploting backends not (yet) supported by Sherpa.

Which backend is used?
======================

When the `sherpa.plot` module is first imported, Sherpa tries to import the
backends installed with Sherpa in the order listed in the
``options.plot_pkg`` setting from the ``sherpa.rc`` startup file.
The first module that imports successfully is set as the active
backend. The following command prints the name and the location
on disk of that module::

   >>> from sherpa import plot
   >>> print(plot.backend)

> 

Reviewer 3: So, with `pylab` selected and `matplotlib` installed I get

```
In [4]: print(plot.backend)
<sherpa.plot.pylab_backend.PylabBackend object at 0x7f09139257b0>
```

which doesn't quite match the text (or at least, it's not what I expected from "the name and the location on disk" of that module). I think this is likely a holdover from the old behavior.

If I've selected a plot package which is not installed (e.g. `pylab` when matplotlbi is not installed, or `bokeh` when we haven't yet merged that backend ;-) I get

```
In [2]: print(plot.backend)
<sherpa.plot.backends.BasicBackend object at 0x7f0840de1ab0>
```

Author: Good question. What do we want to display? All backends have a "name" attribute, that defaults to the class name ("BasicBackend" in this case) except for pylab, where the name is "pylab" because that's what it was before and I don't want to break the API. It's used, e.g. in ciao_contrib. I could implement `__str__` and return the name instead of the usual `<sherpa.plot.backends.BasicBackend object at 0x7f0840de1ab0>` instead.

Change the backend
------------------

.. warning::
   When the plotting backend is changed, new plotting objects must be
   initialized because objects like e.g. `sherpa.plot.DataPlot` keep a
   copy of all default parameters of the backend that was active when they
   are initialized. A different backend might not understand all those
   parameters.

   In particular, that chnging the backend might not work for plot functions
   in Sherpa's UI layer such as `sherpa.astro.plot.plot_data` because those
   work with objects underneath that are not updated when the plotting backend
   is changed.

> 

Reviewer 3: There's a typo here - `changing` - but I think the text is not clear. It also needs to reference something like `sherpa.astro.ui.plot_data` instead of `sherpa.astro.plot.plot_data` but I can't guarantee that'll work.

After the initial import, the backend can be changed by loading one of
the plotting backends shipped with sherpa (or any other module that
provides the same interface)::

  >>> import sherpa.plot.pylab_backend
  >>> plot.backend = sherpa.plot.pylab_backend

> 

Reviewer 3: This doesn't seem to match now. I have matplotlib installed but the plot_pkg is set to `dummy`:

```
$ grep '^plot_pkg' ~/.sherpa-standalone.rc
plot_pkg   : dummy
```

which gives this

```
In [1]: from sherpa import plot

In [2]: print(plot.backend)
<sherpa.plot.backends.BasicBackend object at 0x7f899740c460>

In [3]: import sherpa.plot.pylab_backend

In [4]: plot.backend = sherpa.plot.pylab_backend

In [5]: print(plot.backend)
<module 'sherpa.plot.pylab_backend' from '/home/dburke/miniconda3/envs/sherpa-pr/lib/python3.10/site-packages/sherpa/plot/pylab_backend.py'>
```

Note that this is now referring to a module, not a backend class.

If I switch to picking up pylab automatically we get the class (as shown above but included here):

```
% grep '^plot_pkg' ~/.sherpa-standalone.rc
plot_pkg   : pylab
```

and

```
In [1]: from sherpa import plot

In [2]: print(plot.backend)
<sherpa.plot.pylab_backend.PylabBackend object at 0x7fefb981db10>
```

Reviewer 3: I assume this should read something like the following (which will mean that there should be some docs to indicate that backend class to use for a module, although you can find it pretty easily):

```
sherpa.plot.backend = sherpa.plot.pylab_backend.PylabBackend()
```

Reviewer 3: I've been able to have the dummy plot backend configured, start a notebook and have no plots (e.g. plot_data warns me no plot and get_data_plot() returns a table) and then switch to the pylb backend with the above and then new plots work and get_data_plot now contains a svg plot rather than a table.

I think we should look at making it easier to change the backend - have some routine that you can send the setting in as if you were editing the config file and have the system do the rest - but that would be in a follow-on PR.

Author: Sorry, yes, that's old. I do have a local branch to run doctests on the rst files that would have caught this, but there are other things with that branc hthat don't work yet (which is why I don't have a PR with it yet). 

I'll ponder about other ways to change the backend. This is the kind of comment I was expecting; I honestly had not checked yet if the references in the sphinx docs are correct, because I expected references on a higher level. There would be another way to implement that same idea for backends: They could be plugins: https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/
That would actually be a natural way to do this - but I'm I afraid using setuptools or pkg_utils as suggested in the link above will cause further difficulties for our ongoing changes to the installtion process. Maybe I can revisit that or find some middle ground to all backends by their names as opposed to the fully qualified path.

Sherpa also provides a decorator to change the backend for just one plot::

>   >>> from sherpa.plot.backends import TemporaryPlottingBackend

Reviewer 3: ```suggestion
  >>> from sherpa.plot import TemporaryPlottingBackend
```

  >>> from sherpa.plot.pylab_backend import PylabBackend
  >>> with TemporarypPlottingBackend(PylabBackend()):
  ...     plotting code here...

> 

Reviewer 3: Yay, I was able to get this working (I don't have a `plot_pkg` line in `~/.sherpa-standalone.rc` here so the default is the dummy plot backend):

```
In [1]: from sherpa.astro.ui import *
WARNING: imaging routines will not be available,
failed to import sherpa.image.ds9_backend due to
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: load_arrays(1, [1, 2, 3], [10, 30, 20])

In [3]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [4]: from sherpa.plot import TemporaryPlottingBackend

In [5]: from sherpa.plot.pylab_backend import PylabBackend

In [6]: with TemporaryPlottingBackend(PylabBackend()):
   ...:     plot_data()
   ...:

**AT THIS POINT THE MATPLOTLIB WINDOW APPEARED***

In [7]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

```

Reviewer 3: Interesting. I now get the following

```
% cat delme.rc 
[options]
plot_pkg : none
% SHERPARC=delme.rc ipython --matplotlib
...
In [1]: from sherpa.astro.ui import *
WARNING: Plotting backend 'none' not found or dependencies missing. Trying next option.
WARNING: Tried the following backends listed in delme.rc: 
['none']
None of these imported correctly, so using the 'BasicBackend'.
List of backends that have loaded and would be available: ['BaseBackend', 'BasicBackend', 'IndepOnlyBackend', 'pylab', 'PylabErrorArea']
WARNING: Cannot import usable I/O backend.
    If you are using CIAO, this is most likely an error and you should contact the CIAO helpdesk.
    If you are using Standalone Sherpa, please install astropy.

In [2]: load_arrays(1, [1, 2, 3], [10, 30, 20])

In [3]: plot_data()
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [4]: dp = get_data_plot()

In [5]: dp.x, dp.y
Out[5]: (array([1, 2, 3]), array([10, 30, 20]))

In [6]: from sherpa.plot import TemporaryPlottingBackend

In [7]: from sherpa.plot.pylab_backend import PylabBackend

In [8]: with TemporaryPlottingBackend(PylabBackend()):
    ...:     plot_data()
    ...: 
WARNING: Plotting backend not changed because <sherpa.plot.pylab_backend.PylabBackend object at 0x7fae10cdd890> is not a known plotting backend.
List of available plotting backends: ['BaseBackend', 'BasicBackend', 'IndepOnlyBackend', 'pylab', 'PylabErrorArea']
WARNING: <class 'sherpa.plot.backends.BasicBackend'> does not implement line/symbol plotting. No plot will be produced.

In [9]: with TemporaryPlottingBackend("pylab"):
    ...:     plot_data()
    ...: 
``` 

So, we can no longer pass a class to the `TemporaryPlottingBackend` cnotext manager. I wonder if we would ever want this?

- ``'dotdash'`` or ``'-.'`` for a dot-dashed line.

Markers
--------
- ``"None"`` as a string or ``""`` (empty string) means that no marker will be
  shown, 
- ``"."`` shows dots,
-  ``"o"`` shows cicles (filled or unfilled depending on the backend), 
- ``"+"`` shows plus signs, 
- ``"s"`` shows squares.

Additional backend-specific settings
====================================

Most plotting backends accept more than just the backend-independent options
listed above. For example, :term:`matplotlib` allows `many different ways to
specify colors to cover the entire RGB range
<https://matplotlib.org/stable/tutorials/colors/colors.html>`_, such as
``color=(.3, .4, .5, .2)`` or ``color='xkcd:eggshell'``. The Sherpa plotting
methods will pass any value to the underlying backend, but only the options in
the backend independent list are guaranteed to work in every case. For example, 

  >>> from sherpa.data import Data1D
  >>> from sherpa.plot import DataPlot
  >>> d = Data1D('example data', [1, 2, 3], [3, 2, 5])
  >>> dplot = DataPlot()
  >>> dplot.prepare(d)
  >>> dplot.plot(markerfacecolor='xkcd:eggshell')

will succeed with the `~sherpa.plot.pylab_backend.PylabBackend`, but raise an

> error if the active backend is `~sherpa.plot.bokeh_backend.BokehBackend`. In

Reviewer 3: So, the two `~sherpa.xxx.FooBackend` lines get displayed as `FooBackend` (so the `~` works), but not as actual links. I think this is because we need to add stub rst files like we do for other modules - e.g. `docs/plots/utils.rst` and `docs/plots/image.rst` - and then reference them in a `toctree`

the underlying backend module, they are not documented here - see the
documentation of the specific plotting module for details. Also, they will fail and
raise an error if the plotting backend in use does not understand the ``url`` keyword.

In some cases, the Sherpa plotting commands create several visualization
elements at the same time (lines, symbols, error bars, axes, labels). This makes
using Sherpa classes convenient, but it also means that the plotting functions
do not offer options to customize each and every part. In general, the plotting
functions pass color, linestyle etc. to the elements that describes the data
(line, marker) and generate labels or axes grids using default settings. Backend
specific code can be used to change the properties of the current figure after
the Sherpa plotting.

Backend interface
=================

.. note::

   This section is mostly relevant for developers or advanced users who write new
   Sherpa plot classes or new backends.

This section describes the API that all Sherpa backends offer to explain how to
use it and why it was designed this way. See `sherpa.plot.backend.BaseBackend`

> for a complete listing of the calling signature for each function. 

Reviewer 3: The `*Backend` are drawn as code, but not as links here - presumably for the same reason as the `*Backend` cases above https://github.com/sherpa/sherpa/pull/1382/files#r877545827

Other annotations (e.g. text labels) might be added to the API in the future.
For this reason new backends should inherit from
`~sherpa.plot.backend.BasicBackend`. Any function added to the API will be
implemented in `~sherpa.plot.backend.BasicBackend` as a no-op with a
warning to the user like "Feature XYZ is not available in your backend ". That
way, all Sherpa plots can immediately make use of newly added functions without
breaking existing plotting backends; the worst that happens is that not all
annotation will be visible in every backend.

Return values
-------------

Sherpa does not expect a specific return argument from any plotting function,
but they are allowed to have return values if that is helpful for their internal
implementation, e.g. in the matplotlib backend, plotting a line might return a
line object so that error bars plotted later can use ``line.color`` to match the
color of that line.

Creating plots and panels, clearing and overplotting
----------------------------------------------------
At this stage, we keep the existing API for creating plot and panels, for
clearing and overplotting, i.e. each of the plotting functions above accepts

>  the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow

Reviewer 3: ```suggestion
the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow
```

the space causes the line to be drawn indented!

Multi-panels plot can be set with clear_window, set_subplot, set_jointplot

> [copy documentation from CAIO 4.14 pylab_backend.clear_window etc. into here because that describes use and function]

Reviewer 3: Make this a `.. todo::` ?

Interaction with interactive plots in the UI
--------------------------------------------
Each backend has additional functions that are called before, during and after
interactive plots (begin, exception, and end), and for the setup of 
multi-panel plots [those are taken essentially unchanged from the 4.14 version, so can copy from code into specs]

Other methods
--------------

Backends need to have a few methods

- ``as_html_XXX`` (where XXX is a plot type) that are used for interactive
  display in the notebook with ``_repr_html_``.  These functions take a plot
  object and return an html representation as a string.
- ``get_XXX_plot/hist_prefs`` (where XXX is a plot type) which returns a
  dictionary of preferences that is used for displaying this plot.

> - `~sherpa.plot.backend.BasicBackend.get_latex_for_string` to format latex in strings.

Reviewer 3: I am not proud of inventing `get_latex_for_string` so if there's any way to get rid of it I'd cheer!

Author: I think that's a fine method. Well defined purpose, easily tested, I like it!

Example for a modified backend
------------------------------
Since Sherpa backends are defined using inheritance, new backends can be
created to change the appearance of plots in ways that are beyond the
options offered as keyword arguments for the existing backends. As an example,
the `~sherpa.plot.pylab_backend.PylabErrorArea` backend changes the way that
x-errors are visualized in plots from simple error bars to a shaded area.

.. literalinclude:: ../../sherpa/plot/pylab_backend_area.py
   :language: python
   :lines: 36-87

.. todo:: 
  
   Put code example in here that uses it?

Testing backends and ploting code
----------------------------------
Currently, Sherpa does not employ pixel-level tests that compare a generated
image pixel-by-pixel to a reference image. While testing like that guarantees
that any and all changes are found, they are susceptable to failing for 
reasons unrelated to Sherpa, such as minor changes in the default of
:term:`matplotlib`. 

Instead, the plotting tests in Sherpa fall into the following categories:

  - Tests that do not depend on the output of a plotting package. Those tests
    could check e.g. the `prepare` stage of a plotting object.

>     They work with any backend (since Sherpa has `sherpa.plot.backends.DummyBackend`

Reviewer 3: I couldn't find `DummyBackend` - has it been renamed to `IndepOnlyBackend`?

      def test_this_plotting_feature(all_plot_backends):
          ... test code here ...

    This fixture will execute the test several times with different active backends.
    Alternatively, sherpa also has a fixture that runs all backends *except* for the 
    `sherpa.plot.backends.IndepOnlyBackend`: `plot_backends`. This is useful if a test
    makes use of options that are not on the Backend-independent list and needs to avoid the
    extra warning that `sherpa.plot.backends.IndepOnlyBackend` would emit.
  - Tests for a specific backend usung code such as ``assert plt.gca().xlabel == "text")``.
    These tests can either use the "require" decorators from `sherpa.utils.testing` or, if parts
    of the test are useful for other backends as well, but skip specific statements if the wrong
    backend is active::

      def test_something_that_also_uses_matplotlib(all_plot_backends):
          ... set up data ...
          pl = DataStuffPlot()
          pl.prepare(data)
          out = plt.plot(color='k')

          assert out.x == [1, 2, 3]

          if plot.backend.name == 'pylab':
              assert f'<summary>{summary} Plot</summary>' in r
          else:

>                 assert f'<summary>{summary} Data (' in r

Reviewer 3: ```suggestion
            assert f'<summary>{summary} Data (' in r
```

************************
Plotting backend details
************************

.. warning::

>    The plotting backend API is currently re-designed and details might change in the near future.

Reviewer 3: ```suggestion
   The plotting backend API should be considered experimental for 4.16.0.
```

Author: I'm working through those comments and "resolve conversation" for those that I've addressed locally. In same cases, I'm not taking the exact wording suggested in review (e.g. here I prefer "currently" instead of "4.16.0" just in case we will consider it experimental in 4.16.1 so we don't have to update that comment every time.)

      >>> print(backend.name)

    However, ``backend`` is now a reference to the backend that
    was active when the import was done. If the backend is changed
    later, ``backend`` will still refer to the old backend. Thus,
    always use ``plot.backend`` to access the active backend.

Change the backend
------------------

.. warning::
   When the plotting backend is changed, new plotting objects must be
   initialized because objects like e.g. `sherpa.plot.DataPlot` keep a
   copy of all default parameters of the backend that was active when they
   are initialized. A different backend might not understand all those
   parameters.

   In particular, Sherpa's UI layer (which contains functions such as
   `sherpa.astro.ui.plot_data`) keeps a reference to specific plot objects.
   Changing the backend by itself will not change those references, and thus
   those function might not work with a new backend.

After the initial import, the backend can be changed by loading one of
the plotting backends shipped with sherpa (or any other module that
provides the same interface)::

  >>> from sherpa.plot import set_backend
  >>> set_backend('pylab')

> Sherpa also provides a decorator to change the backend for just one plot::

Reviewer 3: ```suggestion
Sherpa also provides a context manager to change the backend for just one plot::
```

warning to the user like "Feature XYZ is not available in your backend". That
way, all Sherpa plots can immediately make use of newly added functions without
breaking existing plotting backends; the worst that happens is that not all
annotation will be visible in every backend.

Return values
-------------

Sherpa does not expect a specific return argument from any plotting function,
but they are allowed to have return values if that is helpful for their internal
implementation, e.g. in the `~sherpa.plot.pylab_backend.PylabBackend` backend, plotting a line might return a
line object so that error bars plotted later can use ``line.color`` to match the
color of that line.

Creating plots and panels, clearing and overplotting
----------------------------------------------------
.. todo::
   Add details

Each of the plotting functions above accepts
the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow

Multi-panels plot can be set with ``set_subplot`` and ``set_jointplot``.
See `sherpa.plot.backends.BaseBackend` for a listing of the calling signature.

Interaction with interactive plots in the UI
--------------------------------------------

> Each backend also acts as a context manager: All plotting commands
> in the UI are wrapped in a ``with`` statement like this::

Reviewer 3: ```suggestion
Each backend also acts as a context manager: all plotting commands
in the UI can be wrapped in a ``with`` statement like this::
```

Is the change "are" to "can be" correct here?

    plotting test with every possible backend can make the runtime for the tests long.
    In practice, we thus do not run every test that we could for all backends.
    To run a test for all backends, add the `sherpa.conftest.all_plot_backends`
    `pytest fixture <https://docs.pytest.org/en/stable/explanation/fixtures.html>`_::

      def test_this_plotting_feature(all_plot_backends):
          ... test code here ...

    This fixture will execute the test several times with different active backends.
    Alternatively, sherpa also has a fixture that runs all backends *except* for the
    `sherpa.plot.backends.IndepOnlyBackend`: `plot_backends`. This is useful if a test
    makes use of options that are not on the Backend-independent list and needs to avoid the
    extra warning that `sherpa.plot.backends.IndepOnlyBackend` would emit.
  - Tests for a specific backend using code such as ``assert plt.gca().xlabel == "text")``.
    These tests can either use the "require" decorators from `sherpa.utils.testing` or, if parts
    of the test are useful for other backends as well, but skip specific statements if the wrong
    backend is active::

>     >>> import numpy as np

Reviewer 3: When I `make html` in `docs/` I get the following error:

```
/lagado.real/sherpa/sherpa-main/docs/plots/backends.rst:379: ERROR: Inconsistent literal block quoting.
```

but I'm not at all sure what the problem is

Author: It's caused by something that looks almost like, but not quite, a code block further up.

************************
Plotting backend details
************************

.. warning::

>    The plotting backend API is currently re-designed and should currently be considered

Reviewer 3: ```suggestion
   The plotting backend API is currently being re-designed and should be considered
```

The first module that imports successfully is set as the active
backend. The following command prints the name and the location
on disk of that module::

   >>> from sherpa import plot
   >>> print(plot.backend.name)

.. warning::
    Of course, you could be tempted to write::

      >>> from sherpa.plot import backend
      >>> print(backend.name)

    However, ``backend`` is now a reference to the backend that
    was active when the import was done. If the backend is changed
    later, ``backend`` will still refer to the old backend. Thus,
    always use ``plot.backend`` to access the active backend.

Change the backend
------------------

.. warning::
   When the plotting backend is changed, new plotting objects must be
   initialized because objects like e.g. `sherpa.plot.DataPlot` keep a
   copy of all default parameters of the backend that was active when they
   are initialized. A different backend might not understand all those
   parameters.

   In particular, Sherpa's UI layer (which contains functions such as
   `sherpa.astro.ui.plot_data`) keeps a reference to specific plot objects.
   Changing the backend by itself will not change those references, and thus

>    those function might not work with a new backend.

Reviewer 3: If they've used the `set-plot_backend` from the `Session` class then this should only be a problem for any objects stored by the user (e.g. after `get_data_plot`), as - I hope - `plot_data` will pick up the newly-created object.

But I haven't yet tested that.

After the initial import, the backend can be changed by loading one of
the plotting backends shipped with sherpa (or any other module that
provides the same interface)::

  >>> from sherpa.plot import set_backend
  >>> set_backend('pylab')

Sherpa also provides a context manager to change the backend for just one plot::

  >>> from sherpa.plot import TemporaryPlottingBackend
  >>> with TemporarypPlottingBackend('pylab'):
  ...     x = [1,2,3,4]
  ...     # do some plotting with x

The list of available plotting backend names and the classes that implement them is::

  >>> from sherpa.plot.backends import PLOT_BACKENDS

>   >>> print(PLOT_BACKENDS)

Reviewer 3: Would it be better to acknowledge that `PLOT_BACKENDS` is a dict-like object and instead say the following?

```suggestion
  >>> print(PLOT_BACKENDS.keys())
```

- ``'dotdash'`` or ``'-.'`` for a dot-dashed line.

Markers
--------
- ``"None"`` as a string or ``""`` (empty string) means that no marker will be
  shown,
- ``"."`` shows dots,
-  ``"o"`` shows circles (filled or unfilled depending on the backend),
- ``"+"`` shows plus signs,
- ``"s"`` shows squares.

Additional backend-specific settings
====================================

Most plotting backends accept more than just the backend-independent options
listed above. For example, :term:`matplotlib` allows `many different ways to
specify colors to cover the entire RGB range
<https://matplotlib.org/stable/tutorials/colors/colors.html>`_, such as
``color=(.3, .4, .5, .2)`` or ``color='xkcd:eggshell'``. The Sherpa plotting
methods will pass any value to the underlying backend::

  >>> from sherpa.data import Data1D
  >>> from sherpa.plot import DataPlot
  >>> d = Data1D('example data', [1, 2, 3], [3, 2, 5])
  >>> dplot = DataPlot()
  >>> dplot.prepare(d)
  >>> dplot.plot(markerfacecolor='xkcd:eggshell')

will succeed with the `~sherpa.plot.pylab_backend.PylabBackend`
(activated with ``sherpa.plot.set_backend("pylab")``), but raise an

> error if the active backend is `~sherpa.plot.bokeh_backend.BokehBackend`. In

Reviewer 3: Unfortunately, at the moment there is no `sherpa.plot.bokeh_backend` module. In fact, the remaining text in this paragraph assumes that we have a bokeh backend.

I am willing to let this pass now (as re-writing this would be a lot of wasted work once we get a bokeh backend) but we should check again just before the 4.16.0 code freeze.

Reviewer 3: [also hoping we get a bokeh backend soon so then we don't need to worry]

Author: I have a pretty complete draft. The reason I've not push that yet is that it's a pain to deal with rebases of this branch, which it's build on top of. However, with @DougBurke saying it's close, I'll push the bokeh backend shortly. I do want to get both in.

For this reason new backends should inherit from
`~sherpa.plot.backend.BasicBackend`. Any function added to the API will be
implemented in `~sherpa.plot.backend.BasicBackend` as a no-op with a
warning to the user like "Feature XYZ is not available in your backend". That
way, all Sherpa plots can immediately make use of newly added functions without
breaking existing plotting backends; the worst that happens is that not all
annotation will be visible in every backend.

Return values
-------------

Sherpa does not expect a specific return argument from any plotting function,
but they are allowed to have return values if that is helpful for their internal
implementation, e.g. in the `~sherpa.plot.pylab_backend.PylabBackend` backend, plotting a line might return a
line object so that error bars plotted later can use ``line.color`` to match the
color of that line.

Creating plots and panels, clearing and overplotting
----------------------------------------------------
.. todo::
   Add details

Each of the plotting functions above accepts
the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow

> Multi-panels plot can be set with ``set_subplot`` and ``set_jointplot``.
> See `sherpa.plot.backends.BaseBackend` for a listing of the calling signature.

Reviewer 3: A possibility:

```suggestion
Multi-panels plot can be set with `~sherpa.plot.backends.BaseBackend.set_subplot``
and `~sherpa.plot.backends.BaseBacken.set_jointplot` (from `sherpa.plot.backends.BaseBackend`).
```

An advantage is that the HTML viewer can easily follow the link, and I kept the link to the class since we want it easy for the user to follow what class to use as a base.

Of course, given #1868 we may be reworking this section, but it's not obvious to me we'll get that done for 4.16.0.

    ...     plotobj.plot()

That allows the backend to execute any finishing code after the plots are done,
e.g. to save the plot to disk or to display it in a window.

Other methods
--------------

Backends need to have a few more methods:

- ``as_html_XXX`` (where XXX is a plot type) that are used for interactive
  display in the notebook with ``_repr_html_``.  These functions take a plot
  object and return a html representation as a string.
- ``get_XXX_plot/hist_prefs`` (where XXX is a plot type) which returns a
  dictionary of preferences that is used for displaying this plot.
- `~sherpa.plot.backend.BasicBackend.get_latex_for_string` to format latex in strings.
- ``colorlist(n)`` generates a list of n distinct colors. In backends that only have a
  limited number of colors available, the list might repeat.

How do I implement a new backend?
=================================

Example for a modified backend
------------------------------
Since Sherpa backends are defined using inheritance, new backends can be
created to change the appearance of plots in ways that are beyond the
options offered as keyword arguments for the existing backends. As an example,

> the `~sherpa.plot.pylab_backend.PylabErrorArea` backend changes the way that

Reviewer 3: ```suggestion
the `~sherpa.plot.pylab_area_backend.PylabErrorArea` backend changes the way that
```

x-errors are visualized in plots from simple error bars to a shaded area.

.. literalinclude:: ../../sherpa/plot/pylab_area_backend.py
   :language: python
   :lines: 38-99

.. todo::

   Put code example in here that uses it?

Testing backends and plotting code
----------------------------------
Currently, Sherpa does not employ pixel-level tests that compare a generated
image pixel-by-pixel to a reference image. While testing like that guarantees
that any and all changes are found, they are susceptible to failing for
reasons unrelated to Sherpa, such as minor changes in the default of
:term:`matplotlib`.

> Instead, the plotting tests in Sherpa fall into the following categories:

Reviewer 3: Not for this PR, and I don't want to derail this dicussion so I'll mention it and then move on, but I wonder if we should move backend-specific tests into their own test file. Then all the tests could be easily gated for the presence of a backend, and make it easier to see what checks we make for a particular backend.

There are many downsides, including

- sometimes you want to check with different backends that may have some slightly-different test/code but is testing the same thing
- you have to distinguish between generic "just check we work with a backend" vs "test out this particular behavior with a given backend"
- we have a lot of tests scattered around and cleaning this up is a lot of work for little obvious gain

Author: There are a few backend specific tests (see e.g. `test_pylab_specific.py`) but most tests that we have apply 90% of the test code to all text backends. So, yes, we need a better way to account for the differences, but we don't want to have multiple big files files that have nearly identical tests in them.

  - Tests that do not depend on the output of a plotting package. Those tests
    could check e.g. the ``prepare`` stage of a plotting object.
    They work with any backend (since Sherpa has `sherpa.plot.backends.BasicBackend`
    there is always a backend). For those tests, no special
    setup is needed and they will just be run with whatever backend is loaded

>     based on the installed packages and the `.sherpa.rc` file.

Reviewer 3: ```suggestion
    based on the installed packages and the ``.sherpa.rc`` file.
```

With a single backtick it seemed to lose the leading "." in the HTML sphinx generated. There is a long-standing desire by me to have a "good" way to say "configuration file" - with an appropriate link - but hat's for another year.
